book/src/internals/type-system.md

```diff
@@ -0,0 +1,224 @@
+# Type system
+
+> For variables, the type system determines the allowed values of that term.
+>
+> -- Wikipedia
+
+## Purpose
+
+Each of the SQL DBMSs has their own type system. Thanks to SQL standard, they
+are very similar, but have key differences. For example, SQLite does not have a
+type for date or time or timestamps, but it has functions for handling date and
+time that take ISO8603 strings or ints representing Unix timestamps. So it does
+support most of what is possible to do with dates, it just stores a different
+data format and uses different functions to achieve that.
+
+PRQL's task is to define common description of _data formats_, just as how it
+already defines _data transformations_.
+
+I believe this should best be done in two steps:
+
+1. Define PRQL's Type System (PTS), following principles we think a relational
+   language should have (and not focus on what existing SQL DBMSs have).
+
+2. Define a mapping from the SQL's Type System (STS) into PTS, for each of the
```

Comment:
Since it's not going to be a bijection, shouldn't the mapping be from PTS to STS? The image of this mapping may be a strict subset of the co-domain (i.e. STS) but that doesn't matter since we only care about mapping PRQL to SQL at the moment.

(We discussed that it would be nice to map SQL back to PRQL but is that in scope at the moment?)

Reply:
Thinking about it, I realize we actually need two mappings:

- STS to PTS, for when people (or LSP) want to describe their database in PRQL,
- PTS to STS, for compiler to generate valid SQL for the defined PRQL types.

As you said, image of the second mapping must be a subset of the target STS, so we can translate the queries to SQL. (it will take quite a lot of effort to achieve this, we will start with a working impl for basic types and a few dialects).